### PR TITLE
Resolve conflicts by recomputation

### PR DESCRIPTION
**Pull request by naoyam**
Stacked on top of #3611 

This PR resolves the conflicts found by the analysis added at #3611 by recomputing slice/pad input tensors. With this, fusions like `ResizeSchedulerTest.SliceRotateCatResidual` can be scheduled as a single kernel by the resize scheduler.

Recomputation is not the only possible way to resolve conflicts. We could do, for example, cache a block of a input tensor such that multiple uses of the input can be done by just using the block. This would look more like a produce-based scheduling approach. I prototyped that approach [here](#2869), but it didn't perform well for RoPE.

### CODE REVIEW

**Comment by naoyam** [comment]
!test

**Comment by naoyam** [comment]
!test

**Comment by naoyam** [comment]
!test

File: csrc/scheduler/resize.cpp [hunk 1]
```diff
@@ -133,6 +120,30 @@ bool ResizeScheduler::canScheduleCompileTime(Fusion* fusion) {
     return false;
   }
 
+  for (auto out_tv : ir_utils::filterByType<TensorView>(fusion->outputs())) {
```
Comment (naoyam) [inline csrc/scheduler/resize.cpp:123]: This check is needed since the non-exclusivity check is dropped. It was redundant before.

**Comment by naoyam** [comment]
Found a bug. Will update soon.

**Comment by naoyam** [comment]
!test

File: csrc/scheduler/resize.cpp [hunk 2]
```diff
@@ -133,6 +120,44 @@ bool ResizeScheduler::canScheduleCompileTime(Fusion* fusion) {
...
+      continue;
+    }
+    auto exprs = ValGraphBFS::getExprGroupsBetween(
+                     broadcast_graph,
+                     broadcast_graph.toGroups(ref_tv->getLogicalDomain()),
+                     broadcast_graph.toGroups(out_tv->getLogicalDomain()),
+                     /*require_all_to_visited=*/false)
+                     .first;
+    for (const auto& [expr_g, dir] : exprs) {
+      if (expr_g->front()->isA<Resize>()) {
```
Comment (jacobhinkle) [inline csrc/scheduler/resize.cpp:148]: Is my understanding correct that we will segment if we have one resized output and one not resized?
```
addInput(tv0)
tv1 = 2 * tv0;
tv2 = slice(tv1);
tv3 = 3 * tv2;
addOutput(tv2);
addOutput(tv3);
```
In that case we will still have a resize between the `tv2` and `tv3` but it seems like we would potentially be able to schedule it like `tv3`.

**Review by jacobhinkle — APPROVED** [review]
LGTM

File: csrc/scheduler/resize.cpp (hunk 2 above)
Comment (naoyam) [reply csrc/scheduler/resize.cpp:148]: Yes, it is definitely possible, but it's a non-trivial problem to pick a reference tensor. If the dimension of `tv2` is not that different from `tv1`, either of `tv2` or `tv3` should be fine. However, if if they are significantly different, it's unclear if we should fuse them or not. For a trivial fusion like this, it's definitely better to fuse. For more complex fusions, we may be able to generate more efficient segmented kernels. At this point, this is not something I'm trying to address.

**Comment by naoyam** [comment]
!build

**Comment by naoyam** [comment]
!build

**Pull request merged**